### ISSUE TITLE
Exclusion zone endpoints

### ISSUE THREAD

**Issue by mexxexx** [TDAmeritrade/stumpy#131]
Right now, when applying the exclusion zone for self joins, the code is as follows:

```
zone_start = max(0, trivial_idx - excl_zone)
zone_stop = min(k, trivial_idx + excl_zone)
D[zone_start : zone_stop] = np.inf
```
Since slicing works with half open intervals `D[zone_stop]` is not set to `np.inf`. This makes the exclusion zone asymmetric. We should have 'if x is in the exclusion zone of y, then y is in the exclusion zone of x', but this is not the case now:

``` 
0 1 2 3 4 5 6 7 8 9
x x x x
    x x x x
```
In this example `excl_zone=2` and the upper row shows the exclusion zone of `trivial_idx=2` and the lower one the zone of `trivial_idx=4`. 

The issue with this is that the two lowest points in the Matrix Profile have to be the nearest neighbors of each other (the top motif). But right now it might be possible, that this is not the case (In our example): `4` might be the nearest neighbor of `2`, and `P[2]` might be the lowest value in the Matrix Profile. Then `I[2]=4` and we should have `I[4]=2` and `P[4]=P[2]`. But this cannot be the case, as 2 is in the exclusion zone of 4. (`P` is the Matrix Profile and `I` are the Matrix Profile indices)

A fix is to have `D[zone_start : zone_stop+1] = np.inf` instead.

I know that this is a very minor issue and a case that is highly unlikely to occur in real world applications, but it was a bug I came across when writing tests so maybe it should be fixed.

**Comment by seanlaw** [comment]
When you say:

> Since slicing works with half open intervals D[zone_stop] is not set to np.inf

Do you mean that `D[zone_start : zone_stop] = np.inf` is inclusive of `D[zone_start]` and exclusive `D[zone_end]`

I struggled with this when I made this slicing decision initially since, depending on the width of the window `m`, the size of the exclusion zone `(-m/4, m/4)` relative to the trivial index can be a floating point value. I could not come up with a satisfactory solution and so I opted for consistency.

So, whatever happens (I think your proposal is acceptable), the changes need to be consistent throughout the code base (including tests, GPU, and multi-dimensional matrix profile generation scenarios).

**Comment by seanlaw** [comment]
It's minor but I prefer this instead:

```
zone_stop = min(k, trivial_idx + excl_zone + 1)
D[zone_start : zone_stop] = np.inf
```

**Comment by mexxexx** [comment]
> Do you mean that D[zone_start : zone_stop] = np.inf is inclusive of D[zone_start] and exclusive D[zone_end]

Yes exactly.

> I struggled with this when I made this slicing decision initially since, depending on the width of the window m, the size of the exclusion zone (-m/4, m/4) relative to the trivial index can be a floating point value. I could not come up with a satisfactory solution and so I opted for consistency.

I see this point, but this is handled when you compute the `excl_zone=int(np.ceil(m / 4))`. So the exclusion zone always is an integer. But left and right from the trivial index the same number of values should be masked, currently it is one less on the right side, because `D[zone_end]` is excluded.

**Comment by seanlaw** [comment]
I'll preface my comment below by saying that I don't think that `D[zone_start : zone_stop] = np.inf` was the best choice but I wanted to share it for perspective/transparency/open discussion:

I reached back into my memory bank and eventually remembered the original, pragmatic rationale for choosing `D[zone_start : zone_stop] = np.inf`. In the absence of any discussion with anybody else, the simple reason for this choice was that it tried follow the original paper and make the total length of the exclusion zone `m/4 + m/4 = m/2`. In the PR, the length of the exclusion zone becomes `1 + m/2` since the trivial index itself is not included in the length. I actually considered this in the original code but, frankly, I think that the PR solution is more intuitive. It's subtle but I prefer the symmetry as well.

**Comment by seanlaw** [comment]
Fixed in PR #132
